libs/sdk-ui-dashboard/src/model/state/_infra/objRefMap.ts

```diff
@@ -0,0 +1,201 @@
+            this.itemsByIdentifier[`${this.idRefToKey(idExtract(i), this.config.type)}`] = i;
+            this.items.push([refExtract(i), i]);
+            this.size++;
+        });
+
+        return this;
+    }
+
+    public [Symbol.iterator](): IterableIterator<[ObjRef, T]> {
+        return this.items[Symbol.iterator]();
+    }
+
+    public clear(): void {
+        this.items = [];
+        this.itemsByIdentifier = {};
+        this.itemsByUri = {};
+        this.size = 0;
+    }
+
+    public entries(): IterableIterator<[ObjRef, T]> {
+        return this.items[Symbol.iterator]();
+    }
+
+    public get(key: ObjRef): T | undefined {
+        if (isIdentifierRef(key)) {
+            return this.itemsByIdentifier[this.idRefToKey(key.identifier, key.type)];
+        }
+
+        return this.itemsByUri[key.uri];
+    }
+
+    public keys(): IterableIterator<ObjRef> {
+        return this.items.map((i) => i[0])[Symbol.iterator]();
+    }
+
+    public values(): IterableIterator<T> {
+        return this.items.map((i) => i[1])[Symbol.iterator]();
+    }
+}
+
+const metadataObjectExtractors = {
+    idExtract: (i: IMetadataObject) => i.id,
+    uriExtract: (i: IMetadataObject) => i.uri,
+    refExtract: (i: IMetadataObject) => i.ref,
+};
+
+/**
+ * Creates {@link ObjRefMap} for catalog attribute items. Either normal attributes or catalog date attributes.
+ *
+ * @param items - items to add into mapping
+ * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
+ */
+export function newCatalogAttributeMap(
+    items: ReadonlyArray<ICatalogAttribute | ICatalogDateAttribute>,
+    strictTypeCheck: boolean,
+): ObjRefMap<ICatalogAttribute | ICatalogDateAttribute> {
+    const map = new ObjRefMap<ICatalogAttribute | ICatalogDateAttribute>({
+        type: "attribute",
+        strictTypeCheck,
+        idExtract: (i) => i.attribute.id,
+        uriExtract: (i) => i.attribute.uri,
+        refExtract: (i) => i.attribute.ref,
+    });
+
+    return map.fromItems(items);
+}
+
+/**
+ * Creates {@link ObjRefMap} for attribute display form metadata objects.
+ *
+ * @param items - items to add into map
+ * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
+ */
+export function newDisplayFormMap(
+    items: ReadonlyArray<IAttributeDisplayFormMetadataObject>,
+    strictTypeCheck: boolean,
+): ObjRefMap<IAttributeDisplayFormMetadataObject> {
+    const map = new ObjRefMap<IAttributeDisplayFormMetadataObject>({
+        type: "displayForm",
+        strictTypeCheck,
+        ...metadataObjectExtractors,
+    });
+
+    return map.fromItems(items);
+}
+
+/**
+ * Creates {@link ObjRefMap} for attribute metadata objects.
+ *
+ * @param items - items to add into map
+ * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
+ */
+export function newAttributeMap(
+    items: IAttributeMetadataObject[],
+    strictTypeCheck: boolean,
```

Comment:
Should we make a `newAttributeMapStrict` instead of this boolean parameter?

Reply:
Or is this to be set dynamically based on some backend capability?

Reply:
fixed